src/panoptes/pocs/focuser/birger.py

```diff
@@ -249,36 +253,32 @@ def _send_command(self, command, response_length=None):
         # Clear the input buffer in case there's anything left over in there.
         self._serial_port.reset_input_buffer()
 
-        # Send command
-        self._serial_io.write(command + '\r')
-
-        # In verbose mode adaptor will first echo the command
-        echo = self._serial_io.readline().rstrip()
-        assert echo == command, self.logger.warning("echo != command: {} != {}".format(
-            echo, command))
-
-        # Adaptor should then send 'OK', even if there was an error.
-        ok = self._serial_io.readline().rstrip()
-        assert ok == 'OK'
-
         # Depending on which command was sent there may or may not be any further
         # response.
         response = []
 
-        if response_length == 0:
-            # Not expecting any further response. Should check the buffer anyway in case an error
-            # message has been sent.
-            if self._serial_port.in_waiting:
-                response.append(self._serial_io.readline())
+        for i in range(self._max_command_attempts):
+            # Send the command
+            self._serial_io.write(command + '\r')
 
-        elif response_length > 0:
-            # Expecting some number of lines of response. Attempt to read that many lines.
-            for i in range(response_length):
-                response.append(self._serial_io.readline())
+            # In verbose mode adaptor will first echo the command
+            echo = self._serial_io.readline().rstrip()
 
-        else:
-            # Don't know what to expect. Call readlines() to get whatever is there.
-            response.append(self._serial_io.readlines())
+            # Adaptor should then send 'OK', even if there was an error.
+            ok = self._serial_io.readline().rstrip()
+
+            if echo == command and ok == 'OK':
+                # Get any avaialble response left in buffer.
+                response.append(self._serial_io.readlines())
+
+                self.logger.debug(f"Got correct response after {i + 1} attempts")
+                break
+
+            if echo != command:
+                self.logger.warning(f"echo != command: {echo!r} != {command!r}. Retrying command.")
+
+            if ok != 'OK':
+                self.logger.warning(f"ok != 'OK': {ok!r} != 'OK'. Retrying command.")
```

Comment:
You are checking the echo and the 'OK' inside the loop, but you've left the check for an error message outside the loop. I know that our bad serial connections are (so far) causing things to fail at the echo check but it is possible that other intermittent Birger problems could pass the echo and 'OK' check but cause occasional error messages to occur (e.g. ERR21) and the retry code could work around those too if you move the check inside the loop.